### ISSUE TITLE
Aggregator for ConfigMap resource implementation

### ISSUE THREAD

**Issue by pmorie** [kubernetes/kubernetes#18455]
This issue aggregates the PRs for the implementation of the config resource described in #6477

- [X] Proposal #6477
- [x] API Resource #6245  
- [x] Kubectl tooling and associated tests #18712 
- [x] Environment variables and associated E2E tests #18974
- [x] Move to main API #19716
- [x] API resource integration tests #19792 
- [x] Make namespace controller delete ConfigMaps #20645
- [x] Volume Source and associated E2E tests #20114
    - [x] Generalize atomic projections into volumes #20397
- [x] Add kubectl create command #20777
- [x] Documentation #19782
- [x] Quota for ConfigMap #22183 

cc @tmrts

**Comment by tmrts** [comment]
We'll need usage docs as well. A lot of people are anxious to try it out.
- [ ] User Documentation

**Comment by therc** [comment]
The issue numbers for "Move to main API" and "Documentation" are the same; is that intended? I wanted to point out that https://github.com/kubernetes/kubernetes/blob/master/docs/design/configmap.md still talks about [valueFrom: configMap: configMapName:], while it should be [valueFrom: configMapKeyRef: name:]. It's not clear where to report that, though.

**Comment by thockin** [comment]
@pmorie

**Comment by pmorie** [comment]
@therc updated, thanks for noticing that!

**Comment by therc** [comment]
Is there any plan to allow consuming ConfigMap items as e.g. command line flags? Even in #18215 I don't see any mentions of CM (the conversation is hard to follow...). And the downward API only sets environment variables. All I want to do is set e.g. the cluster name or some other cluster-wide constant once, then have a bunch of specs reflect that ("--domain=CLUSTER.example.com"). I'm not even talking about conditionals, regular expressions, arbitrary math leading to computing Mandelbrot sets, etc.: just straight text substitutions as a starting point.

**Comment by thockin** [comment]
You can already do env-var expansion in commandlines, so you should be able
to define an env var from a ConfigMap, then reference that env var in your
commandline as $(NAME)

**Comment by pmorie** [comment]
> You can already do env-var expansion in commandlines, so you should be able to define an env var from a ConfigMap, then reference that env var in your commandline as $(NAME)

bingo

**Comment by mwielgus** [comment]
cc: @piosz @fgrzadkowski

**Comment by davidopp** [comment]
@pmorie Please be sure to explain this prominently in the documentation (a lot of people will have the same question @therc had, since that's almost certainly going to be the main way that people consume ConfigMap).

**Comment by davidopp** [comment]
@pmorie All of the check boxes are checked now :) Should we close the issue?

**Comment by yvespp** [comment]
When I do "kubectl get" I don't see configmap in the error message, shouldn't it be there?
```
lodur:config yves$ kubectl version
Client Version: version.Info{Major:"1", Minor:"2+", GitVersion:"v1.2.0-alpha.8", GitCommit:"5918967a86489460846af94a5c393d971337be29", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"2+", GitVersion:"v1.2.0-alpha.8", GitCommit:"5918967a86489460846af94a5c393d971337be29", GitTreeState:"clean"}
lodur:config yves$ kubectl get 

You must specify the type of resource to get. Valid resource types include:
   * componentstatuses (aka 'cs')
   * events (aka 'ev')
   * endpoints (aka 'ep')
   * horizontalpodautoscalers (aka 'hpa')
   * ingress (aka 'ing')
   * jobs
   * limitranges (aka 'limits')
   * nodes (aka 'no')
   * namespaces (aka 'ns')
   * pods (aka 'po')
   * persistentvolumes (aka 'pv')
   * persistentvolumeclaims (aka 'pvc')
   * quota
   * resourcequotas (aka 'quota')
   * replicasets (aka 'rs')
   * replicationcontrollers (aka 'rc')
   * secrets
   * serviceaccounts
   * services (aka 'svc')
error: Required resource not specified.
```

**Comment by pmorie** [comment]
@yvespp that's weird -- i'll look into it.

@davidopp if you don't mind i'd like to keep it open to track and follow-up tasks that shake out over the next day or two.

**Comment by bgrant0607** [comment]
@pmorie Can this be closed? Is there an issue for docs?

**Comment by bgrant0607** [comment]
I see a PR for docs. Closing.
